File: src/services/auth/permissions.ts
// Permissions Service
// Role-based access control

import type { UserRole } from './types'
import { AuthService, getAuthService } from './auth'

export type Permission =
  | 'dashboard:read'
  | 'kernel:read'
  | 'kernel:write'
  | 'kernel:execute'
  | 'kernel:restart'
  | 'trainer:read'
  | 'trainer:write'
  | 'trainer:execute'
  | 'trainer:pause'
  | 'trainer:resume'
  | 'replay:read'
  | 'replay:write'
  | 'replay:execute'
  | 'plugins:read'
  | 'plugins:write'
  | 'plugins:execute'
  | 'plugins:manage'
  | 'ai:read'
  | 'ai:write'
  | 'ai:execute'
  | 'metrics:read'
  | 'metrics:write'
  | 'metrics:export'
  | 'metrics:reset'
  | 'logs:read'
  | 'logs:export'
  | 'logs:clear'
  | 'commands:read'
  | 'commands:execute'
  | 'commands:queue'
  | 'commands:cancel'
  | 'framework:read'
  | 'framework:shutdown'
  | 'framework:restart'
  | 'license:read'
  | 'license:write'
  | 'license:manage'
  | 'settings:read'
  | 'settings:write'
  | 'users:read'
  | 'users:write'
  | 'users:manage'
  | 'audit:read'

// Permission matrix
const ROLE_PERMISSIONS: Record<UserRole, Permission[]> = {
  [UserRole.VIEWER]: [
    'dashboard:read',
    'kernel:read',
    'trainer:read',
    'replay:read',
    'plugins:read',
    'ai:read',
    'metrics:read',
    'logs:read',
    'commands:read',
    'framework:read',
    'license:read',
    'settings:read',
  ],
  [UserRole.OPERATOR]: [
    // All viewer permissions
    'dashboard:read',
    'kernel:read',
    'trainer:read',
    'replay:read',
    'plugins:read',
    'ai:read',
    'metrics:read',
    'logs:read',
    'commands:read',
    'framework:read',
    'license:read',
    'settings:read',
    // Operator permissions
    'kernel:execute',
    'trainer:execute',
    'trainer:pause',
    'trainer:resume',
    'replay:execute',
    'commands:execute',
    'commands:queue',
  ],
  [UserRole.DEVELOPER]: [
    // All operator permissions
    'dashboard:read',
    'kernel:read',
    'kernel:execute',
    'trainer:read',
    'trainer:execute',
    'trainer:pause',
    'trainer:resume',
    'replay:read',
    'replay:execute',
    'plugins:read',
    'ai:read',
    'metrics:read',
    'logs:read',
    'commands:read',
    'commands:execute',
    'commands:queue',
    'framework:read',
    'license:read',
    'settings:read',
    // Developer permissions
    'kernel:write',
    'kernel:restart',
    'trainer:write',
    'replay:write',
    'plugins:write',
    'plugins:execute',
    'plugins:manage',
    'ai:write',
    'ai:execute',
    'metrics:write',
    'metrics:export',
    'metrics:reset',
    'logs:export',
    'commands:cancel',
  ],
  [UserRole.ADMINISTRATOR]: [
    // All developer permissions
    'dashboard:read',
    'kernel:read',
    'kernel:write',
    'kernel:execute',
    'kernel:restart',
    'trainer:read',
    'trainer:write',
    'trainer:execute',
    'trainer:pause',
    'trainer:resume',
    'replay:read',
    'replay:write',
    'replay:execute',
    'plugins:read',
    'plugins:write',
    'plugins:execute',
    'plugins:manage',
    'ai:read',
    'ai:write',
    'ai:execute',
    'metrics:read',
    'metrics:write',
    'metrics:export',
    'metrics:reset',
    'logs:read',
    'logs:export',
    'logs:clear',
    'commands:read',
    'commands:execute',
    'commands:queue',
    'commands:cancel',
    'framework:read',
    'license:read',
    'license:write',
    'license:manage',
    'settings:read',
    'settings:write',
    'users:read',
    'users:write',
    'audit:read',
    // Administrator permissions
    'framework:shutdown',
    'framework:restart',
  ],
  [UserRole.OWNER]: [
    // All permissions
    'dashboard:read',
    'kernel:read',
    'kernel:write',
    'kernel:execute',
    'kernel:restart',
    'trainer:read',
    'trainer:write',
    'trainer:execute',
    'trainer:pause',
    'trainer:resume',
    'replay:read',
    'replay:write',
    'replay:execute',
    'plugins:read',
    'plugins:write',
    'plugins:execute',
    'plugins:manage',
    'ai:read',
    'ai:write',
    'ai:execute',
    'metrics:read',
    'metrics:write',
    'metrics:export',
    'metrics:reset',
    'logs:read',
    'logs:export',
    'logs:clear',
    'commands:read',
    'commands:execute',
    'commands:queue',
    'commands:cancel',
    'framework:read',
    'framework:shutdown',
    'framework:restart',
    'license:read',
    'license:write',
    'license:manage',
    'settings:read',
    'settings:write',
    'users:read',
    'users:write',
    'users:manage',
    'audit:read',
  ],
}

export class PermissionsService {
  private authService: AuthService

  constructor(authService?: AuthService) {
    this.authService = authService || getAuthService()
  }

  // Check if current user has permission
  hasPermission(permission: Permission): boolean {
    const state = this.authService.getState()
    if (!state.user) return false

    const permissions = ROLE_PERMISSIONS[state.user.role]
    return permissions.includes(permission)
  }

  // Check if user has any of the permissions
  hasAnyPermission(permissions: Permission[]): boolean {
    return permissions.some((p) => this.hasPermission(p))
  }

  // Check if user has all of the permissions
  hasAllPermissions(permissions: Permission[]): boolean {
    return permissions.every((p) => this.hasPermission(p))
  }

  // Get all permissions for current user
  getPermissions(): Permission[] {
    const state = this.authService.getState()
    if (!state.user) return []

    return [...ROLE_PERMISSIONS[state.user.role]]
  }

  // Get minimum role for permission
  getMinimumRoleForPermission(permission: Permission): UserRole | null {
    for (const [role, permissions] of Object.entries(ROLE_PERMISSIONS)) {
      if (permissions.includes(permission)) {
        return role as UserRole
      }
    }
    return null
  }

  // Check if role can access resource
  canAccessResource(role: UserRole, resource: string): boolean {
    const permissions = ROLE_PERMISSIONS[role]
    return permissions.some((p) => p.startsWith(resource.split(':')[0]))
  }

  // Get role display name
  static getRoleDisplayName(role: UserRole): string {
    const names: Record<UserRole, string> = {
      [UserRole.VIEWER]: 'Viewer',
      [UserRole.OPERATOR]: 'Operator',
      [UserRole.DEVELOPER]: 'Developer',
      [UserRole.ADMINISTRATOR]: 'Administrator',
      [UserRole.OWNER]: 'Owner',
    }
    return names[role] || 'Unknown'
  }

  // Get role description
  static getRoleDescription(role: UserRole): string {
    const descriptions: Record<UserRole, string> = {
      [UserRole.VIEWER]: 'Read-only access to dashboard',
      [UserRole.OPERATOR]: 'Can control training and replay',
      [UserRole.DEVELOPER]: 'Can manage plugins and export data',
      [UserRole.ADMINISTRATOR]: 'Full system control except user management',
      [UserRole.OWNER]: 'Full access to all features',
    }
    return descriptions[role] || ''
  }

  // Get all roles
  static getAllRoles(): UserRole[] {
    return [
      UserRole.VIEWER,
      UserRole.OPERATOR,
      UserRole.DEVELOPER,
      UserRole.ADMINISTRATOR,
      UserRole.OWNER,
    ]
  }
}

// Singleton instance
let permissionsInstance: PermissionsService | null = null

export function getPermissionsService(): PermissionsService {
  if (!permissionsInstance) {
    permissionsInstance = new PermissionsService()
  }
  return permissionsInstance
}

File: dashboard_gateway/auth/permissions.py
"""
Permissions
===========
Role-based access control (RBAC).
"""

from enum import Enum
from typing import Dict, List, Set


class Role(str, Enum):
    """User roles."""
    VIEWER = "viewer"
    OPERATOR = "operator"
    DEVELOPER = "developer"
    ADMINISTRATOR = "administrator"
    OWNER = "owner"


# Permission definitions
class Permission(str, Enum):
    # Dashboard
    DASHBOARD_READ = "dashboard:read"
    
    # Kernel
    KERNEL_READ = "kernel:read"
    KERNEL_WRITE = "kernel:write"
    KERNEL_EXECUTE = "kernel:execute"
    KERNEL_RESTART = "kernel:restart"
    
    # Trainer
    TRAINER_READ = "trainer:read"
    TRAINER_WRITE = "trainer:write"
    TRAINER_EXECUTE = "trainer:execute"
    TRAINER_PAUSE = "trainer:pause"
    TRAINER_RESUME = "trainer:resume"
    
    # Replay
    REPLAY_READ = "replay:read"
    REPLAY_WRITE = "replay:write"
    REPLAY_EXECUTE = "replay:execute"
    
    # Plugins
    PLUGINS_READ = "plugins:read"
    PLUGINS_WRITE = "plugins:write"
    PLUGINS_EXECUTE = "plugins:execute"
    PLUGINS_MANAGE = "plugins:manage"
    
    # AI
    AI_READ = "ai:read"
    AI_WRITE = "ai:write"
    AI_EXECUTE = "ai:execute"
    
    # Metrics
    METRICS_READ = "metrics:read"
    METRICS_WRITE = "metrics:write"
    METRICS_EXPORT = "metrics:export"
    METRICS_RESET = "metrics:reset"
    
    # Logs
    LOGS_READ = "logs:read"
    LOGS_EXPORT = "logs:export"
    LOGS_CLEAR = "logs:clear"
    
    # Commands
    COMMANDS_READ = "commands:read"
    COMMANDS_EXECUTE = "commands:execute"
    COMMANDS_QUEUE = "commands:queue"
    COMMANDS_CANCEL = "commands:cancel"
    
    # Framework
    FRAMEWORK_READ = "framework:read"
    FRAMEWORK_SHUTDOWN = "framework:shutdown"
    FRAMEWORK_RESTART = "framework:restart"
    
    # License
    LICENSE_READ = "license:read"
    LICENSE_WRITE = "license:write"
    LICENSE_MANAGE = "license:manage"
    
    # Settings
    SETTINGS_READ = "settings:read"
    SETTINGS_WRITE = "settings:write"
    
    # Users
    USERS_READ = "users:read"
    USERS_WRITE = "users:write"
    USERS_MANAGE = "users:manage"
    
    # Audit
    AUDIT_READ = "audit:read"


# Role permissions matrix
ROLE_PERMISSIONS: Dict[Role, Set[Permission]] = {
    Role.VIEWER: {
        Permission.DASHBOARD_READ,
        Permission.KERNEL_READ,
        Permission.TRAINER_READ,
        Permission.REPLAY_READ,
        Permission.PLUGINS_READ,
        Permission.AI_READ,
        Permission.METRICS_READ,
        Permission.LOGS_READ,
        Permission.COMMANDS_READ,
        Permission.FRAMEWORK_READ,
        Permission.LICENSE_READ,
        Permission.SETTINGS_READ,
    },
    Role.OPERATOR: {
        # Viewer permissions
        Permission.DASHBOARD_READ,
        Permission.KERNEL_READ,
        Permission.TRAINER_READ,
        Permission.REPLAY_READ,
        Permission.PLUGINS_READ,
        Permission.AI_READ,
        Permission.METRICS_READ,
        Permission.LOGS_READ,
        Permission.COMMANDS_READ,
        Permission.FRAMEWORK_READ,
        Permission.LICENSE_READ,
        Permission.SETTINGS_READ,
        # Operator permissions
        Permission.KERNEL_EXECUTE,
        Permission.TRAINER_EXECUTE,
        Permission.TRAINER_PAUSE,
        Permission.TRAINER_RESUME,
        Permission.REPLAY_EXECUTE,
        Permission.COMMANDS_EXECUTE,
        Permission.COMMANDS_QUEUE,
    },
    Role.DEVELOPER: {
        # Operator permissions
        Permission.DASHBOARD_READ,
        Permission.KERNEL_READ,
        Permission.KERNEL_EXECUTE,
        Permission.TRAINER_READ,
        Permission.TRAINER_EXECUTE,
        Permission.TRAINER_PAUSE,
        Permission.TRAINER_RESUME,
        Permission.REPLAY_READ,
        Permission.REPLAY_EXECUTE,
        Permission.PLUGINS_READ,
        Permission.AI_READ,
        Permission.METRICS_READ,
        Permission.LOGS_READ,
        Permission.COMMANDS_READ,
        Permission.COMMANDS_EXECUTE,
        Permission.COMMANDS_QUEUE,
        Permission.FRAMEWORK_READ,
        Permission.LICENSE_READ,
        Permission.SETTINGS_READ,
        # Developer permissions
        Permission.KERNEL_WRITE,
        Permission.KERNEL_RESTART,
        Permission.TRAINER_WRITE,
        Permission.REPLAY_WRITE,
        Permission.PLUGINS_WRITE,
        Permission.PLUGINS_EXECUTE,
        Permission.PLUGINS_MANAGE,
        Permission.AI_WRITE,
        Permission.AI_EXECUTE,
        Permission.METRICS_WRITE,
        Permission.METRICS_EXPORT,
        Permission.METRICS_RESET,
        Permission.LOGS_EXPORT,
        Permission.COMMANDS_CANCEL,
    },
    Role.ADMINISTRATOR: {
        # Developer permissions
        Permission.DASHBOARD_READ,
        Permission.KERNEL_READ,
        Permission.KERNEL_WRITE,
        Permission.KERNEL_EXECUTE,
        Permission.KERNEL_RESTART,
        Permission.TRAINER_READ,
        Permission.TRAINER_WRITE,
        Permission.TRAINER_EXECUTE,
        Permission.TRAINER_PAUSE,
        Permission.TRAINER_RESUME,
        Permission.REPLAY_READ,
        Permission.REPLAY_WRITE,
        Permission.REPLAY_EXECUTE,
        Permission.PLUGINS_READ,
        Permission.PLUGINS_WRITE,
        Permission.PLUGINS_EXECUTE,
        Permission.PLUGINS_MANAGE,
        Permission.AI_READ,
        Permission.AI_WRITE,
        Permission.AI_EXECUTE,
        Permission.METRICS_READ,
        Permission.METRICS_WRITE,
        Permission.METRICS_EXPORT,
        Permission.METRICS_RESET,
        Permission.LOGS_READ,
        Permission.LOGS_EXPORT,
        Permission.LOGS_CLEAR,
        Permission.COMMANDS_READ,
        Permission.COMMANDS_EXECUTE,
        Permission.COMMANDS_QUEUE,
        Permission.COMMANDS_CANCEL,
        Permission.FRAMEWORK_READ,
        Permission.LICENSE_READ,
        Permission.LICENSE_WRITE,
        Permission.LICENSE_MANAGE,
        Permission.SETTINGS_READ,
        Permission.SETTINGS_WRITE,
        Permission.USERS_READ,
        Permission.USERS_WRITE,
        Permission.AUDIT_READ,
        # Administrator permissions
        Permission.FRAMEWORK_SHUTDOWN,
        Permission.FRAMEWORK_RESTART,
    },
    Role.OWNER: set(Permission),  # All permissions
}


class PermissionChecker:
    """
    Permission checker for RBAC.
    """
    
    @staticmethod
    def get_role_level(role: Role) -> int:
        """Get numeric level for a role."""
        levels = {
            Role.VIEWER: 0,
            Role.OPERATOR: 1,
            Role.DEVELOPER: 2,
            Role.ADMINISTRATOR: 3,
            Role.OWNER: 4,
        }
        return levels.get(role, 0)
        
    @staticmethod
    def has_role(minimum_role: Role, user_role: Role) -> bool:
        """Check if user has minimum required role."""
        return PermissionChecker.get_role_level(user_role) >= PermissionChecker.get_role_level(minimum_role)
        
    @staticmethod
    def has_permission(role: Role, permission: Permission) -> bool:
        """Check if role has a specific permission."""
        return permission in ROLE_PERMISSIONS.get(role, set())
        
    @staticmethod
    def has_any_permission(role: Role, permissions: List[Permission]) -> bool:
        """Check if role has any of the specified permissions."""
        role_perms = ROLE_PERMISSIONS.get(role, set())
        return any(p in role_perms for p in permissions)
        
    @staticmethod
    def has_all_permissions(role: Role, permissions: List[Permission]) -> bool:
        """Check if role has all of the specified permissions."""
        role_perms = ROLE_PERMISSIONS.get(role, set())
        return all(p in role_perms for p in permissions)
        
    @staticmethod
    def get_permissions(role: Role) -> List[str]:
        """Get all permissions for a role."""
        return [p.value for p in ROLE_PERMISSIONS.get(role, set())]
        
    @staticmethod
    def get_role_permissions(role: Role) -> Dict[str, List[str]]:
        """Get permissions grouped by resource."""
        permissions = ROLE_PERMISSIONS.get(role, set())
        
        grouped: Dict[str, List[str]] = {}
        for perm in permissions:
            resource = perm.value.split(":")[0]
            action = perm.value.split(":")[1]
            if resource not in grouped:
                grouped[resource] = []
            grouped[resource].append(action)
            
        return grouped
        
    @staticmethod
    def get_all_roles() -> List[Dict]:
        """Get all roles with their permissions."""
        return [
            {
                "name": role.value,
                "level": PermissionChecker.get_role_level(role),
                "permissions": PermissionChecker.get_permissions(role),
            }
            for role in Role
        ]
